### ISSUE TITLE
Does this build on NT 3.51 and/or (DOS-based) 3.1?

### ISSUE THREAD

**Issue by i336** [Microsoft/winfile#9]
The primary way people are going to want to experience this is with period graphics/visual styling. That's primarily going to be DOS-based 3.1 and NT 3.51.

Mostly for the purposes of general documentation, I'm curious to find out how well this would cope with running on

- NT 3.51
- NT 3.1
- DOS-based 3.1
- DOS-based 3.11

and, if there are any issues, if anybody without much inside knowledge would be able to tackle them.

**Comment by craigwi** [comment]
Given the history of WinFile, it probably can be compiled to run on those systems.  If the changes are small, I'd look at a pull request.  Feel free to submit one.

**Comment by DrFrankenstein** [comment]
I doubt that it would have any trouble building/running on NT 3.51, though it would help to have the project file in older formats along with .sln/.vcxproj.

I could give it a shot this week if I find some time.

**Comment by yuhong** [comment]
Win9x would be most interesting though.

**Comment by craigwi** [comment]
As noted, feel free to propose a specific set of changes.  Otherwise, I'll consider this 'won't fix.'

**Comment by NazmusLabs** [comment]
I recall that Win3.1x has the WinG components that can be added on that allows us to run some win32 (32-bit, as opposed to win16 that is native to 3.1) apps in Windows 3.1x. I am curious to see if it "just works" in Win3.1x.

**Comment by ZanderBrown** [comment]
[Looks like it was new in 95/NT4](https://msdn.microsoft.com/en-us/library/windows/desktop/bb776779(v=vs.85).aspx)

**Comment by NazmusLabs** [comment]
@ZanderBrown thanks for the info! I'm hopeful we will eventually have a "retro" branch that would contain as many of the new changes made to v10.0 and beyond without breaking compat. Windows 3.1 probably the best way you can run 16-bit apps on Windows 10 x64 going forward (yes, way better than an XP VM), and having a good file manager becomes important!


P.S. looks like you are running Win3.1 on a very low 16 colors mode, as it's not rendering the default windows cyan theme correctly and, instead, showing a navy blue color. I highly recommend using Win3.1 on DOSBOX, as it can provide sound, true 32-bit color support, and high resolutions beyond the wildest dreams of 1992. It also offers Win3.1 to access NTFS drives and removes any size limits, and you can browse and modify files in your 3.1 environment from File explorer. So yes, you can literally download an image in MS Edge (or whatever), save it, alt+tab to your 3.1 desktop and open the file in a 16 bit app. Impossible to do in a VM.


![windows 3 1 file manger true color on dosbox ss1804](https://user-images.githubusercontent.com/5572377/38567413-28afd7d4-3cb4-11e8-8fb8-e02d8bf627ac.png)


The screenshot you see is my Windows 3.1 environment. Notice the dark theme. Thanks to Windows 3.1's highly customizable UI recolor options, I was able to use the Windows 3.1 control panel to create a modern looking and flat dark theme that mimics the look of Visual studio and the like. The selection highlight and text highlight colors are taken from VS's dark them itself. The wallpaper is there to show that Windows 3.1 is truly running in 32-bit color depth.

**Comment by ZanderBrown** [comment]
Yes the personalisation options of Win3.1 are insane, didn't spend time setting up the display drivers and personally I've always prefered the low colour ¯\_(ツ)_/¯

**Comment by NazmusLabs** [comment]
@ZanderBrown It's not the low color that bothers me in 3.1 (they are ugly in Windows 9x but looks fine in 3.1), but it's the terrible mouse support, causing frustrating mouse lag and frame drops in any VM software. The reason I use DOSBOX is that the mouse moves almost as fluidly as it does on Win10. But, perhaps you are using Win3.1 on DOSBOX, which means you should be getting better mouse support. :)

**Comment by ZanderBrown** [comment]
@NazmusLabs For this I'm running Win3.1 in a kvm so it has a 'real' MSDOS base, for any 'serious' use it's DOSBox

However DOSBox doesn't support SHARE so win32s won't work under it

**Comment by MartinPayne** [comment]
You can [install MS-DOS in DOSBox](https://www.transmissionzero.co.uk/computing/installing-msdos-in-dosbox/) and then install Windows 3.1 and Win32s on that. It will happily run early Win32 applications and I find it to be a lot better than under VirtualBox.

This won't compile for Win16 but I'm sure it could run on Win32s, perhaps with some small modifications. Visual C++ 2.0 can create Win32s applications and can be downloaded from Microsoft if you have a Visual Studio Subscription.

**Comment by NazmusLabs** [comment]
@MartinPayne that's an interesting thought. I didn't think I would be able to install dos on top of DOSBox, as, DOS was an actual operating system whereas Windows 3.1 is just an operating environment running on top of DOS.


Also, I had mentioned earlier that we would need to create a separate retro fork to make it work, as even version 10.0 makes some changes that Win3.1 can't use, such as placing preferences in the App Data folder. We'd need to build off of the original_plus version and then manually back port only the compatible changes from the master branch. That'd be the easiest approach than trying start from a fork of the master branch.

**Comment by MartinPayne** [comment]
@NazmusLabs DOSBox is really just a CPU emulator with enough BIOS and DOS interrupts implemented to get DOS software working. It lets you boot your own code and replace interrupt vectors as you see fit. It's probably best not to start the debate of whether Windows 3.1 was really an operating system, and whether Windows ran on top of DOS or DOS ran on top of Windows. 😂

I'll have a look at the weekend to see whether it can be compiled with MSVC 2.0 and whether it runs on Win32s.

@ZanderBrown The DOSBox commands aren't really that powerful, and the MS-DOS commands significantly exceed their functionality. The main disadvantage is that it's not so easy to share files with the host, but that's equally true of something like VirtualBox. DOSBox also has the advantage that it more closely represents the hardware of the era, for example the OPL3.

**Comment by NazmusLabs** [comment]
@MartinPayne Ah, thanks for the explanations. When I personally say Windows 3.1 runs on top of DOS, I typically talk as an abstract, and not technically or how it's actually programmed. It's easier and more practical to think of it that way because we launch DOS and then run Windows. Sure, behind the scenes it may be more complex, but for normal day to day operations, and especially for normal users who might be trying to get a 16-bit retro Windows 3.1 game running, it's far more practical and efficient to explain to them in this simplified way.

However, what I find funny is that some people calls Windows 3.1 an "app" that runs on DOS, not much more different than, say DOOM. Ouch. And I guess that works to, as it gets the point across for new users interested in retro software or retro gaming.


As for Win32s, I'd love to know how it works out. I don't have an MSDN subscription so I don't have access to the compilers. And to be honest, the thought of even ever needing a 16 bit compiler for a modern application was beyond absurd...until...now. Wow...talk about expecting the unexpected when it comes to software development...

**Comment by DrFrankenstein** [comment]
To answer the title question; _no_, it does not build on Win 3.x, NT or not; or Windows 95 for that matter. It was mostly broken by the port to 64-bit; notably, the use of `INT_PTR`, `LONG_PTR` et al, which don't exist in older SDKs.

I tried with Visual C++ 4.0 and confirmed it just spits out errors all over the place.

16-bit support is even further back, as the qualifiers for x86-16's fancy pointers (`__far`, `__near`, etc.) have long been removed.

As far as __running__ the binaries from the Releases section goes... those were built with a recent toolkit, and support for Windows (NT) 3.x is long gone. Inspecting the image in Dependency Walker reveals some broken dependencies in kernel32.dll, and the VC Redistributables installer won't launch either.
![image](https://user-images.githubusercontent.com/44741/38596924-72cb0582-3d22-11e8-85ec-21c164233038.png)

Of course, if it won't run on WinNT 3.51, it certainly won't run in Win 3.11+Win32s either. And I can't make a special build for those unless I backport this source (note, that's very much doable with a bit of patience).

I hear there's a way to get recent versions of VS to still link against the classic msvcrt.dll, which might get it to work. If anyone is familiar with this, I'd be interested to hear about it.

**Comment by yuhong** [comment]
How about VC6 with a newer SDK?

**Comment by MartinPayne** [comment]
I've given it a go in MSVC 4.2 and found much the same. There are a few errors about functions which don't exist, but most of the errors are to do with data types.

Looking at the source, there is some threading code in there which would rule out being fully functional on Win32s without some additional work (not sure whether the application would crash or handle this gracefully). Probably best to aim for Windows NT 4 and 9x first!

@DrFrankenstein Yes, it is possible to link against MSVCRT.DLL using newer VS (unsupported of course), but you may find problems with the CRT startup code not being compatible with older versions of Windows. You may also have to be careful with the linker options and edit some values in the PE header to get it working. I'm going to stick with MSVC 4.2 for now.

**Comment by NazmusLabs** [comment]
@DrFrankenstein wow, very informative. Definitely looks like it's not going to be an easy port. But thankfully we have the source code; so, it's in the realm of possibility. The project looks to be better suited as a fork project dedicated to rewriting this application using the source code for 16-bit windows and, older 32 bit NT clients.

**Comment by DrFrankenstein** [comment]
If anyone backports this to old versions of Windows, part of the work has been done in #59 (backport to XP). Hat tip to @roytam1.

**Comment by leeter** [comment]
Please make sure this is done into a separate branch, this shouldn't hold back conformance and features in the main codebase because of legacy support.

**Comment by NazmusLabs** [comment]
Yeah, definitely. If we are going to create a branch for older versions, we might as well have the branch focus on XP, 2000, MT, 9x, and 3x, end let the master branch just be focused on Windows 7 (Vista should automatically be compatible, as it's very similar to 7) and up.

On the visual style issue thread, I wondered out loud the need to test the changes in 2000, as MSDN suggests. But this would solve that problem and not hold back the master branch.

I think we should just start with a fork of the Original_plus and start with the XP support and go from there.

**Comment by NazmusLabs** [comment]
@leeter yeah, I was kinda aware of some of Vista's limitation, but definitely not an expert on that area. However, the reason I noted Vista would automatically work is because WinFile is using the most basic of Win32 and COM technologies. It's not using anything near the level that would require the platform features of Windows 7. I may be wrong, though. Perhaps VS2015 and later just doesn't compile to Vista property, but I'd be surprised of that was the case. Vista was still supported when VS2015 and VS2017 was released.

**Comment by leeter** [comment]
@NazmusLabs IIRC the main holdback that complicates vista targeting is the threadpool support changed significantly between vista and 7. I'd have to see if there are places in the codebase where using threadpool would make sense (e.g. don't hammer screws). More than likely most of that is taken care of in the backend of COM anyway. Realistically I think that's jumping waaaay ahead of ourselves for right now as gradual evolution of the codebase is probably the right path.

**Comment by NazmusLabs** [comment]
@leeter thanks for the explanation on the differences between Vista and 7. But yeah, perhaps staring with xp, 2000, and NT compatibility would be the easiest. And once that's verified working, it would be 9x. Since it doesn't work on 9x either, getting 9x working would be a huge step in getting it working for 3.x.

**Comment by yuhong** [comment]
The original Win95 didn't have SHLWAPI.

**Comment by DrFrankenstein** [comment]
That's possible, @leeter. I was only basing that on the fact that this codebase does include `<shlwapi.h>` and link against the corresponding .lib, but if it's not really used, this is even better.

**Comment by leeter** [comment]
@DrFrankenstein it might be used in other spots, I was just noticing that it would be a good replacement for code within. My philosophy has always been to use OS APIs instead of writing my own unless the OS APIs are just absolute 💩 . Less code I have to maintain and ship patches for.

**Comment by NazmusLabs** [comment]
I agree, @DrFrankenstein, that it's best to use Win32s for Windows 3.x. I'm just hoping someone is able to figure out how to get Win32s running properly with DOSBOX. I'm going to need to try their latest DOSBOX build, as the latest stable is from 2010. Hopefully, 8 years worth of updates improves compatibility.

**Comment by ZanderBrown** [comment]
Without booting 'real' DOS the best I've ever found is `fakeshar` aka 'Fake Share' but its likely things will be subtly broken as it doesn't actually implement any of the SHARE functionality just pretends enough that the `if has_share() {` check will pass

Perhaps if @craigwi has some spare time they can fish out the SHARE sources so we could fork that to make it work in DOSBox?

**Comment by MartinPayne** [comment]
@NazmusLabs Is there a reason why you are trying to use Win32s with the DOS which is built into DOSBox? Win32s does work fine in DOSBox if you install a real MS-DOS on it. It runs Freecell, various Windows 95 applications, and my [basic Win32s application](https://github.com/TransmissionZero/Win32s-Example-Application).

**Comment by NazmusLabs** [comment]
@MartinPayne see my posts about why we need DOSBOX rather than real DOS. Using DOSBOX allows me to get 32-bit color depth, NTFS file system support, and the ability to store Windows 3.1 and its apps and data files in a folder in file explorer, allowing me to access 3.1 files from Windows 10 without worrying about mounting Virtual Hard disks. I also don't get the limitation of disk space. Real dos has limited max disk space. With DOSBOX, I can have Win3.1 store a terabyte of stuff, in theory.

P.S. After installing Win32s, I did get an error, but FreeCell does seem to launch, as well as a 32 bit game.

**Comment by MartinPayne** [comment]
I've had a look at this today with Visual C++ 4.2. I've got it to the point where it successfully builds with two warnings. There are some caveats:

  * It crashes on startup (but not before seeing the application flash up very briefly on Windows NT 3.51 and NT 4!).
  * I can only get the Unicode build to work at the moment. There are a lot of errors in the ANSI build, so no Windows 9x support without significant effort.
  * It won't start on Windows NT 3.1 due to requiring a kernel32 function which doesn't exist in this version.

I suspect the crashing is because I deleted some code to get the build to work. The problematic functions are "IsLFNDrive" and "IsNetDrive". I'm not familiar with either of these, so I'll have a look into them and find a more appropriate solution than deleting the code.

**Comment by MartinPayne** [comment]
You're right. They are in the "shell32.lib" import library but not in the headers. Prototyping them gets the build working, but it's still crashing on startup. I'll do some debugging.

**Comment by ZanderBrown** [comment]
@MartinPayne congrats!

How major are the modifications you made?
(e.g. could they merge with master with reasonable preprocessor directives?)

**Comment by MartinPayne** [comment]
@ZanderBrown Thanks!

The [changes are fairly significant](https://github.com/Microsoft/winfile/compare/master...TransmissionZero:master). I doubt they could easily be merged back into master without making it more difficult for everyone to maintain, and my understanding is that the goal of the upstream project is to continue supporting modern Windows without making significant feature changes.

**Comment by ZanderBrown** [comment]
@MartinPayne ah okay

Yes it's probably not worth merging all of that into master as it's primarily looking forward not back

**Comment by MartinPayne** [comment]
I have it running on Windows NT 3.51 now as well:

![WinFile NT 3.51](https://user-images.githubusercontent.com/19162247/38791611-abc03868-4140-11e8-9f29-e1d1ee172fb8.png)

**Comment by NazmusLabs** [comment]
Yeah, if it can be easily added back, then why not. It'd make this easier for everybody.

Plus, if this gets far enough, a retro branch (version 9, as applied to v10) would be nice to have here. That way, people wouldn't have to hunt for forks. Or at the very least, an an officially sanctioned fork that is linked to from the Readme page so people who want to cab easily get to it.

**Comment by MartinPayne** [comment]
@craigwi Thanks! I think that would be possible with some careful `#ifdef`ing. I had to remove some `#include`s for header files which didn't exist in early SDKs, and add preprocessor definitions and function prototypes for the things which were needed from those headers. We'd need to be careful with the VS2017 build to ensure there aren't errors due to the redefinitions.

I'm +1 on a retro branch on here. I've not looked at the new additions to the master branch in detail, so I'm not sure how much effort it would take to get those working with older Windows versions. Even if we can get those working, it might prove difficult to ensure that all new incoming changes don't break the MSVC 4.2 build.

**Comment by ttsirkia** [comment]
@MartinPayne how did you fix it for NT 3.51? I tried your fork and get the same error message as you mentioned earlier.

**Comment by MartinPayne** [comment]
@ttsirkia This was because `WNetRestoreConnectionW` is called `WNetRestoreConnection` on Windows NT 3.51. I updated the code to try both functions.

I've not committed or pushed that change because I need to verify the solution further. For example, is that an ANSI or Unicode function? Also, the usage of it doesn't match the documentation on MSDN because MSDN suggests there is a third argument to this function. I'll do some more investigation at the weekend.

**Comment by roytam1** [comment]
> Mostly for the purposes of general documentation, I'm curious to find out how well this would cope with running on
>    NT 3.51
>    NT 3.1

I think NT 3.5 is missed here.

**Comment by NazmusLabs** [comment]
I'm considering the current implementation of the "Help" menu. The "retro" (aka version 9.0) branch of the application can sport the current help menu whereas the master branch can update the help menu to either direct users to the HTML Help file, a local web page via their default browser, to GitHub documentations, or a combination of these.

The existing help menu tries to launch Winhelp32, which has been deprecated with Windows Vista and has no support in Windows 10. On Windows Vista to 8.1, trying to launch help results in an alert of the deprecation along with an option to download legacy components from the web. In Windows 10, trying to launch help just throws a critical stop error.

Coincidentally, if the retro branch aims to target Windows XP/Server 2003 R2 and earlier, that's also where Winhelp32's full support line is divided, so the choice to use the old help menu on the retro build works perfectly well in the older branch. That said, I can also see additional menu items even in the retro branch's help menu to have links to the GitHub documentations.

**Comment by ZanderBrown** [comment]
Yes have the online support available in all (again keeping codebases as close as possible) and then `#ifdef retro` the classic `hlp` vs newer `chm`

**Comment by MartinPayne** [comment]
@ttsirkia NT 3.51 support has been committed and pushed to my fork now. It turns out `WNetRestoreConnection` is a Unicode function on Windows NT 3.x, and an unimplemented stub on Win32s.

I'll look at getting this running on Windows 10 next, because Windows 9x and NT 3.1 are both going to be tricky!

Note that I'm aware of a forum thread on BetaArchive which says I'm working on a 16 bit version. This is not true (but I'll never say never).

**Comment by ZanderBrown** [comment]
![winfilever](https://user-images.githubusercontent.com/17074021/39099287-32c29408-4670-11e8-9224-fb818745ebe1.png)
Coping from discussions in #104

**Comment by MartinPayne** [comment]
@ZanderBrown Looks mostly fine to me. I would have put all the XPs and Server 2003s on the retro side though because they are out of support. Unless there are plans to ensure that all future changes work in those older OSes, I would expect the red bar to move to the right over time.

**Comment by ZanderBrown** [comment]
@MartinPayne yes we flip flopped over that a few times in #104

**Comment by NazmusLabs** [comment]
@MartinPayne the biggest reason for the line being where it is is because of x64 support. Having server 2003 and xp x64 would meaning needing to compile x64 builds. We thought that the retro branch should only be 32 bit focused.

Secondly, xp is much older than  Server 2003/xp x64, a couple of years in fact. However, my understanding is that in terms of platform level APIs, Server 2003 is closer to XP than it is to Vista.

**Comment by MartinPayne** [comment]
I don't think that would require a 64 bit build, because Windows XP 64 bit should support 32 bit applications, and the other 64 bit Windows certainly do. But anyway, the diagram looks like a sensible plan and it can always be changed later if supporting them in master proves too difficult.

I see we now have a retro branch, so let's start getting some PRs ready! There are a few things we should decide on about the retro branch:

  1. Which toolsets should be supported? MinGW seems sensible, because that's freely available and supports NT 4 and Windows 95. MSVC 4.2 is a good IDE for this and also supports NT 4 and Windows 95. Should we continue to maintain Visual Studio 2017 project files?
  2. How should it be approached? Is my method of replacing all `GetWindowLongPtr`, `INT_PTR`, etc. with `GetWindowLong` and `BOOL` etc. ok, or do we want to try and get clever with the preprocessor?
  3. The retro branch contains a lot of commits since the "original_plus" tag. How much effort should go into supporting those? Should we just revert any commits which prove problematic to support? If we're not interested in those new commits, should we just hard reset the branch back to "original_plus"?

Let's discuss, then I can look at putting my commits into a PR and submit it.

**Comment by ttsirkia** [comment]
This is just my opinion but how much there really are users who would like to use this with the ancient systems? At least for me, it is fantastic that I can build this with NT 3.51 but I'm not really using it there for my daily work. Most of the people are likely using Win10 and such and therefore it would be quite OK that new features etc. exist only in the master branch?

**Comment by NazmusLabs** [comment]
@ttsirkia the idea is that a different branch for older version means the master branch can continue to focus on the latest features. Versions of Windows from the 90s have become a timeless legacy, especially for retro PC enthusiasts and retro gamers. The 3.x and 9x lines of Windows are the almost universally agreed upon operating systems that almost anyone building or restoring a 90s era hardware will install. They, as an extension of DOS are the PC gaming platform from the 90s. Thus, these operating system continue to remain timeless even in virtual machines. DOS has full support for running 3.11, allowing you to seamlessly run any 16 bit app you might want. Having a good file manager is a great edition for these operating systems and enthusiasts.

@MartinPayne yes, we do have a retro v9 branch now! And the decision was made to branch from the latest commit rather than original_plus because the idea, as discussed, is to keep the code bases as close as practically possible. This leads to a few loose guidelines:

1) if there are multiple ways to implement a new feature in master, go with the one that will cause the least compatibility problems with the OSes retro branch supports.
2) If a change is required to improve compatibility with older OS can be added to master without significant rework, then @craigwi has agreed to consider such changes to be merged back into master to keep the code bases as closely aligned as possible.
3) If an existing features can be reimplimented to increase compatibility on the older OS and such change doesn't hinder the application in the newer OSes, that can be done in both branches.
4) only take out commits that were added to master from the retro branch if it cannot made to work on the retro branch.
5) We will need to document the difference between the two branches as they occur so as to have a good understanding how to deal with future merge conflicts.

One great example of these that has already occured is the Reimplimentation of visual styles in order to improve compatibility with Windows 2000 and earlier.

As for the toolset, we haven't decided on anything yet. I'll definitely have to check out your recommendation.

Oh, I think @craigwi is looking for someone to head the retro branch. I'd talk to him if you are interested.

**Comment by ttsirkia** [comment]
@NazmusLabs I meant that the retro branch is very good idea and I appriciate the that has been done to be able to build it with old operating systems. My point was that is it worth of maintaining to separate branches and port the master branch modifications there or would it be enough for the retro branch that the current (original) version can be compiled by using old IDEs?

**Comment by NazmusLabs** [comment]
Oh, and why not just keep the focus on just getting it to run and not bother with back porting commits from master? Well, then what's the fun in that? 😉. Otherwise, once we get it running, the software become "complete" and dead, not to see another enhancements again. But keeping it open to enhancements allows it to remain "living" with the possibly of it getting update. Anyone can pitch in even after the main work to get it compiled is done. This way, we leave the code as a software project that doesn't end and can always be improved upon.

**Comment by ttsirkia** [comment]
I'm not against anything, just wondering how much extra work it mean to port the changes and what is the actual value for it. Of course, if it is purely fun, why not? :) I think this repository is currently one of the most interesting "legacy" projects.

**Comment by ZanderBrown** [comment]
@MartinPayne It would be good to "try and get clever with the preprocessor" within reason to avoid conflicts with master

**Comment by NazmusLabs** [comment]
@ttsirkia ah. I think the idea here is that this branch is for people who want to put in the extra work for whatever reason, be it to learn, develop skill, meet a certain need in their retro setup, etc. Most people will continue to work on master. Essentially, people who are at the time working on this branch must see some value in what they are doing, otherwise why spend so much effort. Perhaps we can look back later and see this as a great thing...or perhaps we will realize that the effort was futile and it was a horrible idea that crashed and burned. 😛😂😅

**Comment by MartinPayne** [comment]
I agree with @ttsirkia, at least in the short term. We know it can work in NT 3.51 to Windows 10, and we think it could work in Windows 9x once the ANSI build is complete. However, including all the feature updates in the retro branch at this point is just moving the goalposts. I'd prefer to concentrate on getting it running in Windows 95, then focus on merging the features gradually so that we can alter any which break the build, or reject any which are never going to work.

**Comment by NazmusLabs** [comment]
@MartinPayne sure, and it was what I initially suggested as well, and, so, I understand what you are saying. So what do you propse? A PR that to the retro that brings it back to the "vanilla" state?

**Comment by MartinPayne** [comment]
@NazmusLabs Or a force push of the "original_plus" tag to the retro branch.

I've been having a go with MinGW this afternoon to see if I can get the build working with that. I'm not quite there yet, but I have confirmed that applications compiled with MinGW are compatible with Windows NT 3.51. Therefore, it might be best if we focus on MinGW for the time being because it's freely available and wouldn't have any problems with `GetWindowLongPtr`s and `INT_PTR`s.

**Comment by NazmusLabs** [comment]
@MartinPayne relieved to hear MinGW us compatible with NT3.51. 


I was wondering if it would be ideal to do a fore push since that gets rid of every single changes post original_plus, as a lot of things like right-click suppor are essential productivity enhancements. Also things like the about dialog boxes, visual styles support, don't known have compat issues. So just wiping out these changes seem a bit drastic.

**Comment by MartinPayne** [comment]
@NazmusLabs It wouldn't be a case of wiping those commits out because they still exist in master. It would just be holding off from bringing in new features until we've got something working, then we could look at bringing in the features one by one and ensuring it still builds.

Looking at the commit history for the MinGW makefile, it looks like the build has been broken twice due to new features being added. The same would happen to the retro build, which means it would be longer before we reach our goal of Windows 9x and potentially 3.1 support.

**Comment by NazmusLabs** [comment]
In that case, submit a pull request to bring the retro branch even with the work you have done in your classic fork. Then others can review it and what not.

**Comment by craigwi** [comment]
So are we satisfied with the mingw version running on Win NT 3.51 and possibly Win3.1?

**Comment by MartinPayne** [comment]
@craigwi For Windows 95 and NT 3.51 (and 3.50?) onwards, yes. For NT 3.1, that wouldn't work because MinGW executables aren't compatible with NT 3.1. I think the only practical option for NT 3.1 is MSVC 2. @malxau appears to have had some success with that as per TransmissionZero/WinFile-Classic#1. For Win32s, I think MSVC 2 is the only option too.

I've had some success with my fork of WinFile built with MinGW, but it crashes on older OSes when optimisation is enabled, and part of the toolbar doesn't display on NT 3.51. It's proving a little tricky to debug though.

**Comment by DosFreak** [comment]
I've been working on the creation of guides for compiling DOSBox and also improving DOSBox host OS compatibility with older versions of Windows. https://www.vogons.org/viewtopic.php?f=31&t=55706

I can confirm that the latest Mingw (not mingw-w64) for Windows can compile working executables for NT3.50+ for i386. 
I can also confirm that Mingw-64 for i686+: from here https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/8.1.0/threads-win32/dwarf/ can compile a working executable for NT3.51+ (Haven't verified 3.50 yet).
MSVCRT: For NT3.51+ you'll need the MSVCRT version from the Google Drive link from my thread.
For NT3.50 scroll down in the thread and grab the MSVCRT version from the 3.50 post.

Also both mingw and mingw-w64 environments have been zipped up and uploaded to my Google Drive so should be pretty easy to develop right away for any project. I do recommend using Linux with mingw-w64 over using mingw-w64 in Windows since occasionally you'll run into issue where the MSYS tools provided are out of date and have bugs whereas in linux they are just an apt-get away. Also compile times are much faster without have to go through MSYS

I've been checking this thread occasionally, keep up the good work guys!

**Comment by emendelson** [comment]
Here's a DOSBox-based Win 3.1 environment (you have to install your own copy of Win 3.1) that has various kludges for transferring files between guest and host:

http://www.columbia.edu/~em32/win31dosbox.html

I've got a version in testing that uses DOSBox-X instead of DOSBox, but the DOSBox version seems to work well enough.

**Comment by HyabusaBadge** [comment]
For SHARE.EXE compatible operating on DOSBOX,
SHARE.EXE (ROM - DOS 7.1)
FAKESHARE.COM
NOSHARE.COM
there is. I recommend NOSHARE.COM.

**Comment by ZanderBrown** [comment]
> FAKESHARE.COM

Is, of course, `FAKESHAR.COM` (8.3 names) and doesn't do anything but exist

Don't know about `NOSHARE.COM` but by it's name i expect it's quite similar so could hardly be considered compatible

**Comment by segin** [comment]
> @NazmusLabs For this I'm running Win3.1 in a kvm so it has a 'real' MSDOS base, for any 'serious' use it's DOSBox
> 
> However DOSBox doesn't support SHARE so win32s won't work under it

Use VSHARE.386 for Win32s. DOSBox (and DOSBox-staging) support the use of Win32s. 
![image](https://github.com/microsoft/winfile/assets/480709/887b8316-259b-4c8e-8f59-48129eea7c3c)

**Comment by roytam1** [comment]
I do hope if Microsoft can release source of 16bit WinFile from Windows 3.0 version for history preservation.
